### PR TITLE
JAMES-2157 HasMimeType only look at the top level mime type

### PR DESCRIPTION
**Pull request by chibenwa**
It do not try to go down the mime tree which makes it impractical for some use cases like `no subpart should have the content type application/zip`.

I understand that some use case do not need to walk down the entire mime tree (eg when categorizing MDN reports only top level content type is enough) however some over generic use cases implies checking the wall mime tree.

Thus I propose to add a `HasMimeTypeAnySubPart` mailet that complement HasMimeType and does just this.

### CODE REVIEW

File: docs/modules/servers/partials/HasMimeType.adoc [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 This matcher checks if the content type matches.
 
+This matcher do not walk down the mime tree and stops at the top level mime part.
```
Comment (jeantil) [inline docs/modules/servers/partials/HasMimeType.adoc:5]:
```suggestion
This matcher does not walk down the mime tree and stops at the top level mime part.
```

File: docs/modules/servers/partials/HasMimeTypeAnySubPart.adoc [hunk 2]
```diff
@@ -0,0 +1,11 @@
+=== HasMimeTypeAnySubPart
+
+This matcher checks if the content type matches.
+
+This matcher walk down the mime tree and will look up at all the mime parts of this message.
```
Comment (jeantil) [inline docs/modules/servers/partials/HasMimeTypeAnySubPart.adoc:5]:
```suggestion
This matcher walks down the mime tree and will look up at all the mime parts of this message.
```

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeType.java [hunk 3]
```diff
@@ -42,6 +42,8 @@
 /**
  * <p>This matcher checks if the content type matches.</p>
  *
+ * <p>This matcher do not walk down the mime tree and stops at the top level mime part.</p>
```
Comment (jeantil) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeType.java:45]:
```suggestion
 * <p>This matcher does not walk down the mime tree and stops at the top level mime part.</p>
```

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeTypeAnySubPart.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+
+/**
+ * <p>This matcher checks if the content type matches.</p>
+ *
+ * <p>This matcher walk down the mime tree and will look up at all the mime parts of this message.</p>
```
Comment (jeantil) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeTypeAnySubPart.java:52]:
```suggestion
 * <p>This matcher walks down the mime tree and will try to match any of the mime parts of this message.</p>
```

**Review by jeantil — APPROVED** [review]

**Comment by chibenwa** [comment]
(Force pushed to solve conflict)

**Pull request merged**